### PR TITLE
Add storage mirror firestore

### PR DESCRIPTION
**Pull request by i14h**
Add storage mirror firestore source code. Credit goes to @jsdt and @kevinthecheung 

- [x] The new extension's id roughly follows the pattern
      `[firebase-product]-[action]-[noun]`, like `storage-resize-images`
- [x] The following match the extension's id:
  - [x] The folder that contains the extension (`my-new-extension/`)
  - [x] The `name` field of `my-new-extension/functions/package.json`
  - [x] The `name` field of `my-new-extension/extension.yaml`
  - [x] A new entry in the `packages` array of the `lerna.json` file in the
        repository's root
- [x] I've added a `generate-readme` script to
      `my-new-extension/functions/package.json` that matches the other
      extensions in the repository
- [x] I've run the following npm commands at the root of the repository:
  - [x] `npm install`
  - [x] `npm run generate-readmes`
  - [x] `npm run format`

### CODE REVIEW

File: storage-mirror-firestore/POSTINSTALL.md [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+will be separated by `/${param:PREFIXES_SUBCOLLECTION_NAME}/` and the Object itself can be found in 
+the `${param:ITEMS_SUBCOLLECTION_NAME}}` in the Prefix Document representing the parent Prefix.
+
+#### Path Mapping Example
+
+For these Cloud Storage paths, Documents will be generated in the following paths in Firestore.
+
+Configuration Parameters:
+- Firestore Root: `${param:FIRESTORE_ROOT}`
+- Items Subcollection Name: `${param:ITEMS_SUBCOLLECTION_NAME}}`
```
Comment (dackers86) [inline storage-mirror-firestore/POSTINSTALL.md:22]:
```suggestion
- Items Subcollection Name: `${param:ITEMS_SUBCOLLECTION_NAME}`
```

File: storage-mirror-firestore/POSTINSTALL.md [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+| --------------------------------- | ------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- |
+| gs://mybucket/root_photo.jpg      | doc(`${param:FIRESTORE_ROOT}`).collection(`${param:ITEMS_SUBCOLLECTION_NAME}}`).doc('root_photo.jpg')                                                                                                                                                                                             |
+| gs://mybucket/2019/08/1/photo.jpg | doc(`${param:FIRESTORE_ROOT}`).collection(`/${param:PREFIXES_SUBCOLLECTION_NAME}/`).doc('2019') .collection(`/${param:PREFIXES_SUBCOLLECTION_NAME}/`).doc('08') .collection(`/${param:PREFIXES_SUBCOLLECTION_NAME}/`).doc('1') .collection(`${param:ITEMS_SUBCOLLECTION_NAME}}`).doc('photo.jpg') |
```
Comment (dackers86) [inline storage-mirror-firestore/POSTINSTALL.md:28]:
```suggestion
| gs://mybucket/2019/08/1/photo.jpg | doc(`${param:FIRESTORE_ROOT}`).collection(`/${param:PREFIXES_SUBCOLLECTION_NAME}/`).doc('2019') .collection(`/${param:PREFIXES_SUBCOLLECTION_NAME}/`).doc('08') .collection(`/${param:PREFIXES_SUBCOLLECTION_NAME}/`).doc('1') .collection(`${param:ITEMS_SUBCOLLECTION_NAME}`).doc('photo.jpg') |
```

File: storage-mirror-firestore/POSTINSTALL.md [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+- Firestore Root: `${param:FIRESTORE_ROOT}`
+- Items Subcollection Name: `${param:ITEMS_SUBCOLLECTION_NAME}}`
+- Prefixes Subcollection Name: `/${param:PREFIXES_SUBCOLLECTION_NAME}/`
+
+| Cloud Storage Path                | Firestore Document Path                                                                                                                                                                                                                                                                           |
+| --------------------------------- | ------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- |
+| gs://mybucket/root_photo.jpg      | doc(`${param:FIRESTORE_ROOT}`).collection(`${param:ITEMS_SUBCOLLECTION_NAME}}`).doc('root_photo.jpg')                                                                                                                                                                                             |
```
Comment (dackers86) [inline storage-mirror-firestore/POSTINSTALL.md:27]:
```suggestion
| gs://mybucket/root_photo.jpg      | doc(`${param:FIRESTORE_ROOT}`).collection(`${param:ITEMS_SUBCOLLECTION_NAME}`).doc('root_photo.jpg')                                                                                                                                                                                             |
```

File: storage-mirror-firestore/POSTINSTALL.md [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+```
+
+### (Optional) Backfill
+
+If you already have objects in GCS that should be mirrored to Firestore, you can use the backfill tool in the stress_test subdirectory. If you application default credentials are set up, you can just run `npm run-script start -- backfill` from that directory.
+
+### (Optional) Cleaning Up Tombstones
+
+You can find instructions on how to use the Tombstone Cleanup tool to clean up all the Firestore Documents corresponding
+to deleted Objects [here](TODO: Tombstone tool url).
```
Comment (dackers86) [inline storage-mirror-firestore/POSTINSTALL.md:49]: Highlighting TODO

File: storage-mirror-firestore/extension.yaml [hunk 5]
```diff
@@ -0,0 +1,193 @@
...
+    validationErrorMessage: Invalid storage bucket
+    example: my-project-12345.appspot.com
+    required: true
+
+  - param: FIRESTORE_ROOT
+    label: Root Document path
+    description: >
+      What is the path to the collection where documents references should be
+      stored?
+    validationRegex: "^[^/]+/[^/]+(/[^/]+/[^/]+)*$" # Must be a path to a document.
```
Comment (dackers86) [inline storage-mirror-firestore/extension.yaml:137]: Comment is covered by the description
```suggestion
    validationRegex: "^[^/]+/[^/]+(/[^/]+/[^/]+)*$"
```

File: storage-mirror-firestore/POSTINSTALL.md [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+| gs://mybucket/2019/08/1/photo.jpg | doc(`${param:FIRESTORE_ROOT}`).collection(`/${param:PREFIXES_SUBCOLLECTION_NAME}/`).doc('2019') .collection(`/${param:PREFIXES_SUBCOLLECTION_NAME}/`).doc('08') .collection(`/${param:PREFIXES_SUBCOLLECTION_NAME}/`).doc('1') .collection(`${param:ITEMS_SUBCOLLECTION_NAME}}`).doc('photo.jpg') |
+
+#### Firestore Query Example
+
+An example of what a Firestore query on Cloud Storage metadata might look like is the following which queries the
+Cloud Storage for the first 5 items ordered by file size.
+
+```JavaScript
+const itemsSnapshot = await firestore
+            .collection('gcs-mirror/your-bucket.appspot.com/items')
```
Comment (dackers86) [inline storage-mirror-firestore/POSTINSTALL.md:37]: Could autofill here based on the params?

```suggestion
            .collection(`${param:FIRESTORE_ROOT}/${${param:ITEMS_SUBCOLLECTION_NAME}}`)
```

**Review by dackers86 — CHANGES_REQUESTED** [review]

File: storage-mirror-firestore/POSTINSTALL.md [hunk 7]
```diff
@@ -0,0 +1,53 @@
+### See it in action
+
+You can test out this extension right away:
+
+1. Go to your [Cloud Storage dashboard](https://console.firebase.google.com/u/0/project/$%7Bparam:PROJECT_ID%7D/storage).
```
[the remark this answers is not in the document]
Comment (jhuleatt) [reply storage-mirror-firestore/POSTINSTALL.md:5]: seems like a weird encoding issue. should use the `${param:PARAMETER_NAME}` syntax

```suggestion
1. Go to your [Cloud Storage dashboard](https://console.firebase.google.com/u/0/project/${param:PROJECT_ID}/storage).
```

File: storage-mirror-firestore/POSTINSTALL.md [hunk 8]
```diff
@@ -0,0 +1,53 @@
...
+2. Upload a file to Cloud storage.
+3. Go to your [Cloud Firestore dashboard](https://console.firebase.google.com/project/$%7Bparam:PROJECT_ID%7D/database/firestore/data)
```
[the remark this answers is not in the document]
Comment (jhuleatt) [reply storage-mirror-firestore/POSTINSTALL.md:7]: same as https://github.com/FirebaseExtended/experimental-extensions/pull/48#discussion_r622164117, should use the `${param:PARAMETER_NAME}` syntax

File: storage-mirror-firestore/POSTINSTALL.md [hunk 9]
```diff
@@ -0,0 +1,53 @@
...
+4. You can check to see that the corresponding Document was created under `${param:FIRESTORE_ROOT}`.
+
+### Using the extension
+
+Any changes made in Cloud Storage will be reflected in the Firestore mirror. Each delimited path prefix segment in the Object path
+will be separated by `/${param:PREFIXES_SUBCOLLECTION_NAME}/` and the Object itself can be found in 
+the `${param:ITEMS_SUBCOLLECTION_NAME}}` in the Prefix Document representing the parent Prefix.
```
[the remark this answers is not in the document]
Comment (jhuleatt) [reply storage-mirror-firestore/POSTINSTALL.md:14]: same as https://github.com/FirebaseExtended/experimental-extensions/pull/48#discussion_r622164117, should use the `${param:PARAMETER_NAME}` syntax

File: storage-mirror-firestore/POSTINSTALL.md (hunk 3 above)
Comment (jhuleatt) [reply storage-mirror-firestore/POSTINSTALL.md:27]: @dackers86 did this render correctly in the console? Not sure if we've tried a markdown table in postinstall yet

File: storage-mirror-firestore/POSTINSTALL.md (hunk 4 above)
Comment (jhuleatt) [reply storage-mirror-firestore/POSTINSTALL.md:49]: @jsdt FYI

File: storage-mirror-firestore/POSTINSTALL.md (hunk 6 above)
Comment (jhuleatt) [reply storage-mirror-firestore/POSTINSTALL.md:37]: Agreed, although I think there's an extra `${}` around the second param in your suggestion, @dackers86

**Comment by i14h** [comment]
@dackers86 @jhuleatt thanks for the feedback. All addressed and PTAL

**Comment by jhuleatt** [comment]
overriding CLA bot - it got confused by an alternate email for @dackers86

File: storage-mirror-firestore/POSTINSTALL.md (hunk 3 above)
Comment (dackers86) [reply storage-mirror-firestore/POSTINSTALL.md]: @jhuleatt It did render a table, however text seems unaligned...?

It rendered similar to...

image

File: storage-mirror-firestore/POSTINSTALL.md (hunk 6 above)
Comment (dackers86) [reply storage-mirror-firestore/POSTINSTALL.md]: Thanks @jhuleatt Good spot!

File: storage-mirror-firestore/POSTINSTALL.md [hunk 10]
```diff
@@ -0,0 +1,52 @@
...
+            .limit(5)
+```
+
+### (Optional) Backfill
+
+If you already have objects in GCS that should be mirrored to Firestore, you can use the backfill tool in the stress_test subdirectory. If you application default credentials are set up, you can just run `npm run-script start -- backfill` from that directory.
+
+### (Optional) Cleaning Up Tombstones
+
+*TODO*: Implement clean-up tool for tombstone documents
```
Comment (dackers86) [inline storage-mirror-firestore/POSTINSTALL.md:48]: Highlighting TODO

**Review by dackers86 — CHANGES_REQUESTED** [review]

**Review by jhuleatt — APPROVED** [review]

File: storage-mirror-firestore/POSTINSTALL.md (hunk 10 above)
Comment (jhuleatt) [reply storage-mirror-firestore/POSTINSTALL.md:48]: +1, as far as I can tell this script doesn't exist, so consider taking out this section or adding some basic instructions for someone to DIY cleanup

**Review by dackers86 — APPROVED** [review]

**Pull request merged**